### PR TITLE
forwarding command wrapping

### PR DESCRIPTION
**Pull request by zapion**
add forward command for adb wrapper

### CODE REVIEW

**Comment by zapion** [comment]
@askeing r?

File: b2g_util/util/adb_helper.py [hunk 1]
```diff
@@ -101,6 +101,48 @@ def adb_push(cls, source, dest, serial=None):
         return output
 
     @classmethod
+    def adb_forward(cls, command=None, source=None, dest=None, serial=None, port=2828):
```
Comment (askeing) [inline b2g_util/util/adb_helper.py]: `port` no use.

File: b2g_util/util/adb_helper.py [hunk 2]
```diff
@@ -101,6 +101,48 @@ def adb_push(cls, source, dest, serial=None):
...
+        p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+        shell_ret, stderr = p.communicate()
+        # TODO: parse list output and return as serial/items
+        logger.debug('CMD: {0}'.format(cmd))
+        logger.debug('RAW_RET: {0}'.format(shell_ret))
+        if command == 'list':
+            if stderr:
+                logger.debug('RAW_ERR: {0}'.format(stderr))
+            if p.returncode is not 0:
+                if command == 'remove':
```
Comment (askeing) [inline b2g_util/util/adb_helper.py]: L128 (`command=='list'`) is conflict with this line.

File: b2g_util/util/adb_helper.py [hunk 3]
```diff
@@ -101,6 +101,48 @@ def adb_push(cls, source, dest, serial=None):
...
+        else:
+            cmd += "%s %s %s" % (cmd, source, dest)
+        cmd = "%s '%s; echo $?'" % (cmd, command)
+        p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+        shell_ret, stderr = p.communicate()
+        # TODO: parse list output and return as serial/items
+        logger.debug('CMD: {0}'.format(cmd))
+        logger.debug('RAW_RET: {0}'.format(shell_ret))
+        if command == 'list':
+            if stderr:
```
Comment (askeing) [inline b2g_util/util/adb_helper.py]: when having `stderr`, should always put it into log.

File: b2g_util/util/adb_helper.py [hunk 4]
```diff
@@ -101,6 +101,48 @@ def adb_push(cls, source, dest, serial=None):
...
+            cmd = 'adb forward'
+        else:
+            cmd = 'adb -s %s forward' % (serial,)
+        # get returncode from device
+        if command:
+            cmd += " --%s" % (command,)
+            if 'remove' == command:
+                cmd += (" " + dest)
+        else:
+            cmd += "%s %s %s" % (cmd, source, dest)
```
Comment (askeing) [inline b2g_util/util/adb_helper.py]: using `'` should better than `"`

File: b2g_util/util/adb_helper.py (hunk 1 above)
Comment (zapion) [inline b2g_util/util/adb_helper.py]: Yes, I forgot it is mentioned in source & dest already.

File: b2g_util/util/adb_helper.py (hunk 2 above)
Comment (zapion) [inline b2g_util/util/adb_helper.py]: Good catch, fixing in upcoming patch

File: b2g_util/util/adb_helper.py [hunk 5]
```diff
@@ -101,6 +101,51 @@ def adb_push(cls, source, dest, serial=None):
...
+        When using remove command, only dest should be mentioned, e.g. adb_forward('remove', dest="tcp:12345")
+        """
+        if serial is None:
+            cmd = 'adb forward'
+        else:
+            cmd = 'adb -s %s forward' % (serial,)
+        # get returncode from device
+        if command:
+            cmd += " --%s" % (command,)
+            if 'remove' == command:
```
Comment (askeing) [inline b2g_util/util/adb_helper.py:118]: `adb forward --no-rebind <local> <remote>` will fail here.
I am trying to add the commit for it.

File: b2g_util/util/adb_helper.py [hunk 6]
```diff
@@ -101,6 +101,51 @@ def adb_push(cls, source, dest, serial=None):
         return output
 
     @classmethod
+    def adb_forward(cls, command=None, source=None, dest=None, serial=None):
+        """
+        Use port forwarding on device.
+        @return: the stdout and return code (from device) of "adb forward" command. e.g. (stdout, retcode)
+        @raise exception: When return code (from adb command) isn't zero.
+        When using remove command, only dest should be mentioned, e.g. adb_forward('remove', dest="tcp:12345")
```
Comment (askeing) [inline b2g_util/util/adb_helper.py:109]: this line will attach to `exception` part.

File: b2g_util/util/adb_helper.py [hunk 7]
```diff
@@ -101,6 +101,51 @@ def adb_push(cls, source, dest, serial=None):
...
+        if serial is None:
+            cmd = 'adb forward'
+        else:
+            cmd = 'adb -s %s forward' % (serial,)
+        # get returncode from device
+        if command:
+            cmd += " --%s" % (command,)
+            if 'remove' == command:
+                # only get dest for remove, assert error if no dest
+                if not dest:
```
Comment (askeing) [inline b2g_util/util/adb_helper.py:120]: the spec is `forward --list <local>` not remote `dest`.

**Pull request closed without merging**